### PR TITLE
DPC-1642: Pass certs from webapp to API

### PR DESCRIPTION
**Pull request by Sun-Mountain**
### Fixes [DPC-1642](https://jira.cms.gov/browse/DPC-1642)

In tandem [DPC-1438](https://jira.cms.gov/browse/DPC-1438), pass certs (in production) through http requests to the api through the implementer app.

### Proposed Changes

Branched off of [branch](https://github.com/CMSgov/dpc-app/pull/1370) for 1438, I added ssl and certificates to the http requests from the implementer app to the api in the `api_client.rb` file.

**Notes for SRE Team:**
Four environment variables have been added:
- DPC_PORTAL_CERT
- DPC_PORTAL_KEY
- DPC_CA_CERT
- DPC_CA_CERT_PATH

`DPC_PORTAL_CERT` and `DPC_PORTAL_KEY` are for base64 encoded certification and key.

`DPC_CA_CERT` is for the base64 encoded ca certification.

`DPC_CA_CERT_PATH` is the path to that certification necessary for the ssl requirements.

### Change Details

- Added a distinction to only run in the production environment (not test or development) to make local development easier.
- Added the `write_ca_cert.rb` initializer to take the `DPC_CA_CERT` environment variable and create the `ca.crt` file in the app route directory for the ssl requirements.

### Security Implications

- [ ] new software dependencies

- [x] security controls or supporting software altered

To this point, the Implementer Portal and DPC v2 API service have not been deployed to an AWS environment. In local testing the Implementer Portal has been accessing the DPC v2 API service unauthenticated. This PR adds authentication--and encryption--through mutual TLS (mTLS).

In DPC v1, the web application used a special "golden" macaroon to authenticate requests to the API service. This DPC v2 web portal (Implementer Portal) is using mTLS instead, so that it will be using the same tool as DPC v2 will be employing in several other services.

- [ ] new data stored or transmitted

- [x] security checklist is completed for this change

[SIA completed](https://confluence.cms.gov/pages/viewpage.action?pageId=490380420)

- [ ] requires more information or team discussion to evaluate security implications

- [x] no PHI/PII is affected by this change

### Acceptance Validation

Screen Shot 2021-08-06 at 3 12 49 PM

Screen Shot 2021-08-06 at 3 13 00 PM

Screen Shot 2021-09-01 at 11 20 45 AM

### CODE REVIEW

File: dpc-impl/app/services/api_client.rb [hunk 1]
```diff
@@ -92,6 +92,10 @@ def get_request(uri_string)
 
   def http_request(request, uri)
     http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = true
```
Comment (MrBilnon) [inline dpc-impl/app/services/api_client.rb:95]: can you add this? 
`http.verify_mode: OpenSSL::SSL::VERIFY_PEER`

File: dpc-impl/app/services/api_client.rb [hunk 2]
```diff
@@ -93,6 +93,14 @@ def get_request(uri_string)
   def http_request(request, uri)
     http = Net::HTTP.new(uri.host, uri.port)
 
+    if use_ssl?
+      http.use_ssl = true
+      http.verify_mode = OpenSSL::SSL::VERIFY_PEER
+      http.cert = OpenSSL::X509::Certificate.new(Base64.decode64(ENV.fetch('DPC_PORTAL_CERT')))
+      http.ca_file = ENV.fetch('DPC_CA_CERT')
+      http.key = OpenSSL::PKey::RSA.new(Base64.decode64(ENV.fetch('DPC_PORTAL_KEY')))
```
Comment (dhgreene) [inline dpc-impl/app/services/api_client.rb:101]: DPC-1653 and DPC-1654 will address inconsistent certificate formats.

**Review by dhgreene — DISMISSED** [review]
👍 This should be merged in conjunction with #1373 and #1370, after ops-related tickets add env var support.

**Review by jonfulk — APPROVED** [review]
Re-approving after dismissing @dhgreene 's review with a formatting commit. 

> 👍 This should be merged in conjunction with #1373 and #1370, after ops-related tickets add env var support.

**Pull request merged**